modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/sources/kinesis/KinesisSource.scala

```diff
@@ -50,109 +58,183 @@ object KinesisSource {
-                // This can happen near the shard end, e.g. the penultimate batch in the shard, when KCL has already enqueued the final record in the shard to the fs2 queue.
-                // We must not release the semaphore yet, because we are not ready for fs2-aws to checkpoint the end of the shard.
-                // We can safely ignore the exception and move on.
-                Logger[F].warn(
-                  s"Checkpointing failed on a record which was not the last in the shard. Meanwhile, KCL has already enqueued the final record in the shard to the fs2 queue. Ignoring error and instead continue processing towards the shard end"
-                )
-            }
-        }
-    def nack(c: Map[String, KinesisMetadata[F]]): F[Unit] = Applicative[F].unit
+    val empty: Map[String, Checkpointable] = Map.empty
+
+    def ack(c: Map[String, Checkpointable]): F[Unit] =
+      c.toList.parTraverse_ {
+        case (shardId, RecordCheckpointable(extendedSequenceNumber, checkpointer)) =>
+          Logger[F].debug(s"Checkpointing shard $shardId at $extendedSequenceNumber") *>
+            Sync[F]
+              .blocking(
+                checkpointer.checkpoint(extendedSequenceNumber.sequenceNumber, extendedSequenceNumber.subSequenceNumber)
+              )
+              .recoverWith(ignoreShutdownExceptions(shardId))
+        case (shardId, ShardEndCheckpointable(checkpointer, release)) =>
+          Logger[F].debug(s"Checkpointing shard $shardId at SHARD_END") *>
+            Sync[F].blocking(checkpointer.checkpoint()).recoverWith(ignoreShutdownExceptions(shardId)) *>
+            Sync[F].delay(release.countDown)
+      }
+
+    def nack(c: Map[String, Checkpointable]): F[Unit] =
+      Sync[F].unit
   }
 
+  private sealed trait KCLAction
+  private case class ProcessRecords(shardId: String, processRecordsInput: ProcessRecordsInput) extends KCLAction
+  private case class ShardEnd(
+    shardId: String,
+    await: CountDownLatch,
+    shardEndedInput: ShardEndedInput
+  ) extends KCLAction
+  private case class KCLError(t: Throwable) extends KCLAction
+
   private def kinesisStream[F[_]: Async](
     config: KinesisSourceConfig,
     livenessRef: Ref[F, FiniteDuration]
-  ): Stream[F, LowLevelEvents[Map[String, KinesisMetadata[F]]]] =
+  ): Stream[F, Stream[F, LowLevelEvents[Map[String, Checkpointable]]]] =
+    for {
+      kinesisClient <- Stream.resource(mkKinesisClient[F](config.customEndpoint))
+      dynamoClient <- Stream.resource(mkDynamoDbClient[F](config.dynamodbCustomEndpoint))
+      cloudWatchClient <- Stream.resource(mkCloudWatchClient[F](config.cloudwatchCustomEndpoint))
+      queue = new SynchronousQueue[KCLAction]
+      scheduler <- Stream.eval(scheduler(kinesisClient, dynamoClient, cloudWatchClient, config, queue))
+      _ <- Stream.resource(runRecordProcessor[F](scheduler))
+      s <- Stream.emit(pullFromQueue(queue, livenessRef).stream).repeat
```

Comment:
I think it is equally as useful/useless as it was before.  Before and after this PR, the KCL starts running when the wrapper calls `scheduler.run()`, and that equates to being "connected" in `LowLevelSource`.  But it is not yet processing any shards.  Later, the KCL obtains the lease for a shard and starts sending events into the app.

The situation is marginally nicer now, because I have some error handling in the `WorkerStateChangeListener`, which we didn't get with the old fs2 wrapper.  The new error handling means our app should crash if KCL failed to initialize.  But still, we should not entirely trust `isConnectedRef`.